taxcalc/parameters.py

```diff
@@ -1,170 +1,496 @@
-                           " values for the years between {0} and {1}.")
-                    raise ValueError(msg.format(self.current_year, year))
-
-            # Set up the '_X = [a, b,...]' variables as 'X = a'
-            self.set_year(self._current_year)
+        if self.current_year != self.start_year:
+            self.set_year(self.start_year)
+        if not reform:
+            return  # no reform to implement
+        reform_years = sorted(list(reform.keys()))
+        last_reform_year = max(reform_years)
+        if last_reform_year > self.end_year:
+            msg = 'reform provision in year={} > end_year={}'
+            ValueError(msg.format(last_reform_year, self.end_year))
+        for year in reform_years:
+            if year != self.start_year:
+                self.set_year(year)
+            self._update({year: reform[year]})
+        self.set_year(self.start_year)
 
     @property
     def current_year(self):
+        """
+        Current policy parameter year property.
+        """
         return self._current_year
 
     @property
     def start_year(self):
+        """
+        First policy parameter year property.
+        """
         return self._start_year
 
     @property
-    def budget_years(self):
-        return self._budget_years
+    def num_years(self):
+        """
+        Number of policy parameter years property.
+        """
+        return self._num_years
 
-    def increment_year(self):
-        self._current_year += 1
-        self.set_year(self._current_year)
+    @property
+    def end_year(self):
+        """
+        Last policy parameter year property.
+        """
+        return self._end_year
 
-    def set_year(self, yr):
-        for name, vals in self._vals.items():
+    def set_year(self, year):
+        """
+        Set policy parameters to values for specified calendar year.
+
+        Parameters
+        ----------
+        year: int
+            calendar year for which to current_year and parameter values
+
+        Raises
+        ------
+        ValueError:
+            if year is not in [start_year, end_year] range.
+
+        Returns
+        -------
+        nothing: void
+
+        Notes
+        -----
+        To increment the current year, use the following statement::
+
+            ppo.set_year(ppo.current_year + 1)
+
+        where ppo is a policy Parameters object.
+        """
+        if year < self.start_year or year > self.end_year:
+            msg = 'year passed to set_year() must be in [{},{}] range.'
+            raise ValueError(msg.format(self.start_year, self.end_year))
+        self._current_year = year
+        year_zero_indexed = year - self._start_year
+        for name in self._vals:
             arr = getattr(self, name)
-            setattr(self, name[1:], arr[yr - self._start_year])
+            setattr(self, name[1:], arr[year_zero_indexed])
+
+    @classmethod
+    def default_data(cls, metadata=False, start_year=None):
```

Comment:
T.J. @talumbau,  @MattHJensen asked me to download the dropQ repo and try out my ideas.  Here is my report on the run_models code changes (to avoid the calculator factory function) that eliminate the errors you were experiencing:
```
NOTES ON DROPQ-MASTER/EXAMPLE.PY USE WITH REDESIGNED PARAMETERS API
===================================================================

(A) out of the box example.py (with only path changes to find files):

iMac2:dropQ-master mrh$ python example.py
You loaded data for 2008.
Your data have beeen extrapolated to 2013.
You loaded data for 2008.
Your data have beeen extrapolated to 2013.
Modifications are for year 2015 and Parameters are for year 2013.
  Parameters will be advanced to year 2015
Modifications are for year 2015 and Records are for year 2008.
  Records will be advanced to year 2015
Traceback (most recent call last):
  File "example.py", line 13, in <module>
    mY_dec, mX_dec, df_dec, mY_bin, mX_bin, df_bin, fiscal_tots = dropq.run_models(tax_dta, num_years=10, user_mods=user_mods)
  File "/Users/mrh/work/OSPC/dropQ-master/dropq/dropq.py", line 324, in run_models
    calc3 = calculator(params=params3, records=records3, mods=user_mods)
  File "/Users/mrh/work/OSPC/dropQ-master/dropq/../../tax-calculator/taxcalc/calculate.py", line 63, in calculator
    calc = Calculator(params, records)
  File "/Users/mrh/work/OSPC/dropQ-master/dropq/../../tax-calculator/taxcalc/calculate.py", line 95, in __init__
    assert self._params.current_year == self._records.current_year
AssertionError

This is the error T.J. reported; the relevant dropq.run_models calc3 code is:
    # User specified Plans
    params3 = Parameters(start_year=2013)
    # Create a Calculator for the user specified plan
    calc3 = calculator(params=params3, records=records3, mods=user_mods)
    assert calc3.current_year == start_year
    calc3.calc_all()

(B) revise calc3 code to avoid the calculator factory function as follows:

    # User specified Plans
    params3 = Parameters(start_year=2013)
    params3.implement_reform(user_mods)
    params3.set_year(start_year)
    # Create a Calculator for the user specified plan
    calc3 = Calculator(params=params3, records=records3)
    calc3.calc_all()
    assert calc3.current_year == start_year

Also, make the following change to handle default_data now being a static
method, not a global function:

    #all_params = sorted([str(k) for k in taxcalc.parameters.default_data()])
    all_params = sorted([str(k) for k in Parameters.default_data()])

With these two sets of run_models code changes, we have the following results:

iMac2:dropQ-master mrh$ python example.py
You loaded data for 2008.
Your data have beeen extrapolated to 2013.
You loaded data for 2008.
Your data have beeen extrapolated to 2013.
You loaded data for 2008.
Your data have beeen extrapolated to 2015.
('elapsed time for second run: ', 3.204669952392578)
('elapsed time for second run: ', 6.002561807632446)
('elapsed time for second run: ', 5.773066997528076)
('elapsed time for second run: ', 5.514545917510986)
('elapsed time for second run: ', 5.405025959014893)
('elapsed time for second run: ', 5.396327018737793)
('elapsed time for second run: ', 5.584659099578857)
('elapsed time for second run: ', 5.554149866104126)
('elapsed time for second run: ', 5.574904918670654)
('elapsed time for second run: ', 5.687175035476685)

Is this the expected output from example.py?
```